services/CurrentTime/source/CurrentTimeService.cpp

```diff
@@ -45,18 +50,60 @@ void CurrentTimeService::set_event_handler(EventHandler *handler) {
     _current_time_handler = handler;
 }
 
-time_t CurrentTimeService::get_time()
+time_t CurrentTimeService::get_time() const
 {
     time_t epoch_time = time(nullptr);
 
     return epoch_time + _time_offset;
 }
 
-void CurrentTimeService::set_time(time_t host_time)
+void CurrentTimeService::set_time(time_t host_time, uint8_t adjust_reason)
 {
     time_t epoch_time = time(nullptr);
 
+    time_t old_time_offset = _time_offset;
     _time_offset = host_time - epoch_time;
+
+    update_current_time_value(adjust_reason, old_time_offset - _time_offset);
+}
+
+void CurrentTimeService::update_current_time_value(const uint8_t adjust_reason, time_t time_offset_difference) {
+    time_t local_time = get_time();
+
+    struct tm *local_time_tm = localtime(&local_time);
+
+    CurrentTime current_time(local_time_tm);
+
+    current_time.adjust_reason = adjust_reason;
+
+    bool send_update = true;
+    if ((current_time.adjust_reason & (MANUAL_TIME_UPDATE | CHANGE_OF_TIME_ZONE | CHANGE_OF_DST)) == 0) {
+        if (timer.elapsed_time() < 15min) {
+            if (std::abs(time_offset_difference) <= 1) {
+                send_update = false;
+            }
+        }
+    }
+
+    if (send_update) {
+        _ble.gattServer().write(_current_time_char.getValueHandle(),
+                                reinterpret_cast<const uint8_t *>(&current_time),
+                                CURRENT_TIME_CHAR_VALUE_SIZE);
+
+        timer.reset();
+        timer.start();
+    }
+
+    start_periodic_time_update();
+}
+
+void CurrentTimeService::start_periodic_time_update() {
+    if (_event_queue_handle == 0) {
```

Comment:
Can you point me out in the spec when it is required to do a periodic update ?

Reply:
By negation of:
"If the time of the server device is changed because of reference time update, then no notifications shall be sent to the client device within the 15 minutes from the last notification"

Negating this: notification will be sent after 15 minutes.

Reply:
Interesting, I didn't interpreted it that way. I was thinking the notification would not be sent at all.

Reply:
But in such case, the periodic time update should not be periodic and only queued if `send_update ` was false in `update_current_time_value`. It also means that when an update is sent, the pending event should be removed. 

What happens to the reason too ? It doesn't seem to be properly captured by `start_periodic_time_update` and if an update requiring immediate notification is sent while another one is pending, shouldn't it add `EXTERNAL_REFERENCE_TIME_UPDATE` to the reasons ?

Reply:
technically the EXTERNAL_REFERENCE_TIME_UPDATE reason should be set every second and cleared when notification is sent

Reply:
of course the above does not apply to user interaction - so if a user sets it with set time or a remote peer sends a value then whatever reason they have given should be the reason used

Reply:
since in effect the only time a EXTERNAL_REFERENCE_TIME_UPDATE is sent is because of the periodic update, it's OK to just rely on the periodic update function to give the reason as otherwise the reason is already supplied correctly